app/models/journeyDomain/package.scala

```diff
@@ -89,52 +102,141 @@ package object domain {
-        case Some(value) => Right(value)
-        case None        => Left(ReaderError(a, message))
+    private def reader(message: Option[String])(implicit reads: Reads[A]): Read[A] = pages => {
+      val fn: UserAnswers => EitherType[ReaderSuccess[A]] = _.get(a) match {
+        case Some(value) => Right(ReaderSuccess(value, pages.append(a)))
+        case None        => Left(ReaderError(a, pages.append(a), message))
       }
       UserAnswersReader(fn)
     }
 
-    def mandatoryReader(predicate: A => Boolean)(implicit reads: Reads[A]): UserAnswersReader[A] = {
-      val fn: UserAnswers => EitherType[A] = _.get(a) match {
-        case Some(value) if predicate(value) => Right(value)
-        case _                               => Left(ReaderError(a))
-      }
-      UserAnswersReader(fn)
-    }
-
-    def optionalReader(implicit reads: Reads[A]): UserAnswersReader[Option[A]] = {
-      val fn: UserAnswers => EitherType[Option[A]] = ua => Right(ua.get(a))
+    def optionalReader(implicit reads: Reads[A]): Read[Option[A]] = pages => {
+      val fn: UserAnswers => EitherType[ReaderSuccess[Option[A]]] = ua => Right(ReaderSuccess(ua.get(a), pages))
       UserAnswersReader(fn)
     }
   }
 
   implicit class JsArrayGettableAsReaderOps(jsArray: Gettable[JsArray]) {
 
-    def arrayReader(implicit reads: Reads[JsArray]): UserAnswersReader[JsArray] = {
-      val fn: UserAnswers => EitherType[JsArray] = ua => Right(ua.get(jsArray).getOrElse(JsArray()))
+    def arrayReader(implicit reads: Reads[JsArray]): Read[JsArray] = pages => {
+      val fn: UserAnswers => EitherType[ReaderSuccess[JsArray]] =
+        ua => Right(ReaderSuccess(ua.get(jsArray).getOrElse(JsArray()), pages.append(jsArray)))
+
       UserAnswersReader(fn)
     }
 
-    def fieldReader[T](page: Index => Gettable[T])(implicit rds: Reads[T]): UserAnswersReader[Seq[T]] = {
-      val fn: UserAnswers => EitherType[Seq[T]] = ua => {
+    def fieldReader[T](page: Index => Gettable[T])(implicit rds: Reads[T]): Read[Seq[T]] = pages => {
+      val fn: UserAnswers => EitherType[ReaderSuccess[Seq[T]]] = ua => {
         Right {
-          ua.get(jsArray).getOrElse(JsArray()).value.indices.foldLeft[Seq[T]](Nil) {
-            (acc, i) =>
-              ua.get(page(Index(i))) match {
-                case Some(value) => acc :+ value
-                case None        => acc
+          ua.get(jsArray).getOrElse(JsArray()).value.indices.foldLeft[ReaderSuccess[Seq[T]]](ReaderSuccess(Nil, pages)) {
+            case (ReaderSuccess(ts, pages), i) =>
+              val gettable = page(Index(i))
+              ua.get(gettable) match {
+                case Some(t) => ReaderSuccess(ts :+ t, pages.append(gettable))
+                case None    => ReaderSuccess(ts, pages.append(gettable))
               }
           }
         }
       }
       UserAnswersReader(fn)
     }
   }
+
+  type Pages   = Seq[Page]
+  type Read[T] = Pages => UserAnswersReader[T]
```

Comment:
I think this confused me initially, I misstook it to be the play.json.lin.Read thing